packages/workflow/src/internals.ts

```diff
@@ -534,14 +542,116 @@ export class Activator implements ActivationHandler {
     );
   }
 
+  public doUpdate(activation: coresdk.workflow_activation.IDoUpdate): void {
+    const { id: updateId, name, headers, runValidator } = activation;
+    if (!updateId) {
+      throw new TypeError('Missing activation update id');
+    }
+    if (!name) {
+      throw new TypeError('Missing activation update name');
+    }
+    if (!this.updateHandlers.has(name)) {
+      // TODO (dan): Signal is able to handle this situation more gracefully by
+      // using this.bufferedSignals. Should something analogous exist for
+      // Update?
+      this.rejectUpdate(updateId, ApplicationFailure.nonRetryable(`Update has no handler: ${name}`));
+      return;
+    }
+    const execute = composeInterceptors(this.interceptors.inbound, 'handleUpdate', this.updateNextHandler.bind(this));
+
+    const validate = composeInterceptors(
+      this.interceptors.inbound,
+      'validateUpdate',
+      this.validateUpdateNextHandler.bind(this)
+    );
+
+    const makeInput = (): UpdateInput => ({
+      updateId,
+      args: arrayFromPayloads(this.payloadConverter, activation.input),
+      name,
+      headers: headers ?? {},
+    });
+
+    // The implementation below is responsible for upholding, and constrained
+    // by, the following contract:
+    //
+    // 1. During validation, any error must fail the Update; during the Update
+    //    itself, Temporal errors fail the Update whereas other errors fail the
+    //    activation.
+    //
+    // 2. The handler must not see any mutations of the arguments made by the
+    //    validator.
+    //
+    // 3. Any error when decoding/deserializing input must be caught and result
+    //    in rejection of the Update before it is accepted, even if there is no
+    //    validator.
+    //
+    // 4. The initial synchronous portion of the (async) Update handler should
+    //    be executed after the (sync) validator completes such that there is
+    //    minimal opportunity for a different concurrent task to be scheduled
+    //    between them.
+    //
+    // 5. The stack trace view provided in the Temporal UI must not be polluted
+    //    by promises that do not derive from user code. This implies that
+    //    async/await syntax may not be used.
+    //
+    // Note that there is a deliberately unhandled promise rejection below.
+    // These are caught elsewhere and fail the corresponding activation.
+    let input: UpdateInput;
+    try {
+      if (runValidator) {
+        validate(makeInput());
+      }
+      input = makeInput();
+    } catch (error) {
+      this.rejectUpdate(updateId, error);
+      return;
+    }
+    this.acceptUpdate(updateId);
+    untrackPromise(
+      execute(input)
```

Comment:
nit: call `makeInput` here.

Reply:
I don't think we can do that: `makeInput` has to be called inside the `try/catch` above because it may fail. That's one of the requirements listed in the comment:

```
    // 3. Any error when decoding/deserializing input must be caught and result
    //    in rejection of the Update before it is accepted, even if there is no
    //    validator.
```